src/Stratis.Bitcoin.Features.BlockStore.Tests/BlockStoreTests.cs

```diff
@@ -0,0 +1,241 @@
+                    HashPrevBlock = chain.Tip.HashBlock,
+                    Bits = Target.Difficulty1
+                };
+                
+                var chainedHeader = new ChainedHeader(header, header.GetHash(), chain.Tip);
+
+                chain.SetTip(chainedHeader);
+            }
+
+            return chain;
+        }
+
+        private async Task WaitUntilQueueIsEmpty()
+        {
+            while (this.blockStore.BlocksQueueCount != 0)
+                await Task.Delay(100).ConfigureAwait(false);
+            
+            // For very slow environments.
+            await Task.Delay(100).ConfigureAwait(false);
+        }
+
+        /// <summary>Checks that block store tip initializes from the hash of the last saved block.</summary>
+        [Fact]
+        public async Task BlockStoreInitializesTipAtHashBlockAsync()
+        {
+            ChainedHeader initializationHeader = this.chain.Tip.Previous.Previous.Previous;
+            this.repository.BlockHash = initializationHeader.HashBlock;
+
+            await this.blockStore.InitializeAsync().ConfigureAwait(false);
+            Assert.Equal(initializationHeader, this.blockStore.StoreTip);
+        }
+        
+        [Fact]
+        public async Task BlockStoreSavesBatchRightAwayWhenAtConsensusTipAsync()
+        {
+            this.repository.BlockHash = this.chain.Tip.Previous.HashBlock;
+
+            await this.blockStore.InitializeAsync().ConfigureAwait(false);
+
+            this.blockStore.AddToPending(new BlockPair(new Block(), this.chain.Tip));
+
+            await this.WaitUntilQueueIsEmpty().ConfigureAwait(false);
+            
+            Assert.Equal(this.blockStore.StoreTip, this.chain.Tip);
+            Assert.Equal(this.chain.Tip.HashBlock, this.repository.BlockHash);
+        }
+
+        [Fact]
+        public async Task BatchIsSavedAfter5MbLimitReachedAsync()
+        {
+            Block block = Block.Load(Encoders.Hex.DecodeData("000000202f6f6a130549473222411b5c6f54150d63b32aadf10e57f7d563cfc7010000001e28204471ef9ef11acd73543894a96a3044932b85e99889e731322a8ec28a9f9ae9fc56ffff011d0011b40202010000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff2c028027266a24aa21a9ed09154465f26a2a4144739eba3e83b3e9ae6a1f69566eae7dc3747d48f1183779010effffffff0250b5062a0100000023210263ed47e995cbbf1bc560101e3b76c6bdb1b094a185450cea533781ce598ff2b6ac0000000000000000266a24aa21a9ed09154465f26a2a4144739eba3e83b3e9ae6a1f69566eae7dc3747d48f1183779012000000000000000000000000000000000000000000000000000000000000000000000000001000000000101cecd90cd38ac6858c47f2fe9f28145d6e18f9c5abc7ef1a41e2f19e6fe0362580100000000ffffffff0130b48d06000000001976a91405481b7f1d90c5a167a15b00e8af76eb6984ea5988ac0247304402206104c335e4adbb920184957f9f710b09de17d015329fde6807b9d321fd2142db02200b24ad996b4aa4ff103000348b5ad690abfd9fddae546af9e568394ed4a83113012103a65786c1a48d4167aca08cf6eb8eed081e13f45c02dc6000fd8f3bb16242579a00000000"), Network.Main);
```

Comment:
can this string be a const?

Reply:
Since it is used only in one method I've kept it inside it.

Reply:
Yeah I thought so. Would having it as a local const give it some meaning as to why it is that string?

Reply:
Optimally, we would have some common ground that would contain all the help of functions and constants including example blocks like this. So then we would not need to repeat the block definition every time. But this is definitely out of scope of this PR.Ivan told me that there is this suggestion somewhere already, so hopefully we will eventually have that.

Reply:
no need for that its clearly a block under test that is being deserialized.